intermediate-python-testing.md

```diff
@@ -300,11 +302,72 @@ def test_do_a_thing(mocker):
+### Getting started
 
-For more on how to use `mock`, see [the quickstart](https://docs.python.org/3/library/unittest.mock.html#quick-guide) in the `mock` documention, as well as [this excellent tutorial](https://www.toptal.com/python/an-introduction-to-mocking-in-python). Meanwhile, here are a few of our own hard-won lessons from early `mock` use.
+#### `patch()`
+
+`patch()` is the main method that allows you to replace a dependency with a mock. There are two main ways to use this method -- as a decorator or as a context manager.
+
+As a decorator, `@patch()` can be used on a testing class or a single test function. The mock that it creates will live for the duration of the class or function. Once the test (or all the tests in the class, as the case may be) is complete, the mock is cleaned up and any other tests will use the real dependency instead of the mock. For example, as a class decorator:
+
+```python
+@patch('path.to.patched_dependency')
+class SomeClass:
+    def first_test(self, patched_dependency):
+        # do the things in the first test using the mock
+        pass
+
+    def second_test(self, patched_dependency):
+        # do the things in the second test using the mock
+        pass
+```
+
+Or as a method decorator:
+```python
+
+class SomeClass:
+    @patch('path.to.patched_dependency')
+    def first_test(self, patched_dependency):
+        # do the things in the first test using the mock
+        pass
+```
+
+As a context manager, `patch()` allows your mock to be scoped to just part of a function, like this:
+```python
+def first_test(self):
+    patched_dependency.return_value = "this is what I expect"
+    with patch('path.to.patched_dependency') as patched_dependency:
+        # do the things in the first test using the mock
+    pass
+```
 
-#### Mock methods where they're used, not defined.
+#### Useful arguments
+
+Part of a mock object's magic is that it does not necessarily behave like the object it's patching. For example, any attribute you call on your mock pops into existence by virtue of being called, even if it doesn't exist on the real dependency. While this can be helpful at times, mock objects can be more useful if they throw errors, return values, and otherwise behave like the objects, functions, or other dependencies they temporarily replace. 
+
+You can pass certain arguments into the `patch()` method which allow you to set attributes on the mock object that allow you to more closely mimic the behavior of the dependency you want to temporarily replace.
+
+`spec`, `auto_spec`, `set_spec` are arguments you can pass into `patch()` that allow the created mock to have the same attributes as the original dependency.
```

Comment:
A link to mock or pytest-mock documentation on these arguments would be very helpful!